### PR TITLE
Add Radius.Core/terraformConfigs and bicepConfigs resources

### LINKED ISSUE #10615: Terraform/Bicep Settings Lifecycle
### Overview of feature request

Externalize Radius Terraform and Bicep recipe configuration into dedicated settings resources (`Radius.Core/terraformSettings` and `Radius.Core/bicepSettings`), centralize Terraform binary lifecycle, and let platform teams supply Terraform settings exactly as they do today.

**Key changes:**
- New `terraformSettings` resource: `.terraformrc`, backend config, env vars, logging
- New `bicepSettings` resource: registry authentication (BasicAuth, Azure WI, AWS IRSA)
- Installer async pipeline for Terraform binary lifecycle (`rad terraform install|uninstall|status`)
- Environments reference settings resources instead of embedded `recipeConfig`

### Acceptance criteria

- Platform engineers can install/uninstall Terraform via CLI with version pinning
- Settings resources are reusable across multiple environments
- Legacy `recipeConfig` continues working with deprecation warnings
- Sequential installs are queued and processed in order
- Uninstall blocks when Terraform executions are in progress

### Additional context

- Feature spec: https://github.com/radius-project/design-notes/pull/107
- Design notes: https://github.com/radius-project/design-notes/pull/117
- Related issue: https://github.com/radius-project/radius/issues/10615

### Tasks

**Phase 1 - Core Implementation**
- [x] Add REST controllers for `Radius.Core/terraformSettings` and `Radius.Core/bicepSettings` with CRUD operations (#11013)
- [x] Add TypeSpec definitions, regenerate SDKs, update datamodel converters (#11013)
- [x] Implement installer status storage and REST endpoints (install/uninstall/status) (#11014)
- [x] Register installer queue/worker with `MaxOperationConcurrency=1` and async handler (#11014)
- [x] Implement binary lifecycle helpers (mirror downloads, checksum validation, PVC layout, symlink management) (#11014)
- [ ] Add CLI commands: `rad terraform install|uninstall|status` with `--wait` semantics (#11049)
- [ ] Wire `Radius.Core/environments` controller to consume new settings resources
- [ ] Add migration scaffolding (warnings, legacy compatibility checks)
- [ ] Update Helm charts (PVC mounts, remove init-container, installer queue config)

[AB#17385](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/17385)

### PR DESCRIPTION
**Pull request by sylvainsf**
# Description

Adds two new resources in the `Radius.Core` namespace — `terraformConfigs` and `bicepConfigs` — that environments reference by resource ID. Platform teams can now define private registry authentication, Terraform CLI configuration (`provider_installation`, `credentials`, env vars), and Bicep registry authentication once and share it across multiple environments.

This implements a scoped subset of [design-notes#107](https://github.com/radius-project/design-notes/pull/107). Out of scope: Terraform binary lifecycle (`rad terraform install`), backend state stores, installer pipeline. See [docs/architecture/terraform-bicep-config.md](docs/architecture/terraform-bicep-config.md) for the full design.

**What's included**

- New TypeSpec/OpenAPI resources `Radius.Core/terraformConfigs` and `Radius.Core/bicepConfigs` (regenerated SDK).
- New `terraformConfig?` and `bicepConfig?` properties on `Radius.Core/environments`, validated for existence at PUT time.
- Datamodels, bidirectional API↔datamodel converters, REST controllers wired in `pkg/corerp/setup/setup.go`.
- Loader bridge in `pkg/recipes/configloader/environment.go` that resolves the new config resources at recipe execution time and populates `RecipeConfig` consumed by the shared driver.
- New `ProviderInstallation` field on the shared `TerraformConfigProperties` datamodel; the Terraform driver writes a `.terraformrc` to the working dir and points Terraform at it via `TF_CLI_CONFIG_FILE` on both Deploy and Delete. Field is optional and only populated by the Radius.Core path — the legacy `Applications.Core` `recipeConfig` flow is unchanged.
- Unit tests for the `.terraformrc` writer, the env-var setter, and the Delete-path helper.
- Functional test (`Test_TerraformConfig_BicepConfig_Combined`) deploying both configs in a single environment and running a Terraform recipe end-to-end.
- Architecture doc: [docs/architecture/terraform-bicep-config.md](docs/architecture/terraform-bicep-config.md).

**Known follow-ups (filed separately)**
- `bicepConfig.registryAuthentication.{AzureWI,AwsIrsa}` are accepted by the schema but not yet wired in the Bicep driver (only `BasicAuth` flows through today).
- `referencedBy` on the config resources is exposed in the schema but not yet populated.
- Delete protection (409 when a config is referenced by an environment) is not yet implemented.

## Type of change

- This pull request adds or changes features of Radius and has an approved issue (issue link required).

Fixes: #10615

## Contributor checklist
Please verify that the PR meets the following requirements, where applicable:

- An overview of proposed schema changes is included in a linked GitHub issue.
    - [x] Yes 
    - [ ] Not applicable 
- A design document PR is created in the [design-notes repository](https://github.com/radius-project/design-notes/), if new APIs are being introduced.
    - [x] Yes 
    - [ ] Not applicable 
- The design document has been reviewed and approved by Radius maintainers/approvers.
    - [x] Yes 
    - [ ] Not applicable 
- A PR for the [samples repository](https://github.com/radius-project/samples) is created, if existing samples are affected by the changes in this PR.
    - [ ] Yes 
    - [x] Not applicable 
- A PR for the [documentation repository](https://github.com/radius-project/docs) is created, if the changes in this PR affect the documentation or any user facing updates are made.
    - [ ] Yes 
    - [x] Not applicable 
- A PR for the [recipes repository](https://github.com/radius-project/recipes) is created, if existing recipes are affected by the changes in this PR.
    - [ ] Yes 
    - [x] Not applicable

### CODE REVIEW

File: docs/architecture/terraform-bicep-config.md [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+  @doc("Resource ID of a Radius.Core/bicepConfig providing Bicep recipe settings.")
+  bicepConfig?: string;
+}
+```
+
+### TerraformConfig shape
+
+Follows the schema from the [feature spec](https://github.com/radius-project/design-notes/pull/107).
+The `terraformrc` property is a like-for-like representation of the Terraform CLI
+configuration file (`.terraformrc`). The `env` property holds non-sensitive
```
Comment (DariuszPorowski) [inline docs/architecture/terraform-bicep-config.md]: Let's make sure `terraform.rc` is supported as well.

Comment (sylvainsf) [reply docs/architecture/terraform-bicep-config.md]: Same structure just a different filename?

File: docs/architecture/terraform-bicep-config.md [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+2. Platform team deploys an `environment` referencing them by resource ID.
+3. Environment controller validates the referenced config resources exist
+   at PUT time and returns `400 Bad Request` if not.
+4. At recipe execution time, the configuration loader fetches the referenced
+   config resources and populates `recipes.Configuration.RecipeConfig`. The
+   loader bridges the new schema into the existing `RecipeConfig` shape so the
+   shared Terraform and Bicep drivers consume it through the same path the
+   legacy `Applications.Core` `recipeConfig` flow uses.
+5. When `terraformrc.providerInstallation` is set, the Terraform driver
+   renders a `.terraformrc` in the working directory and points Terraform at
```
Comment (DariuszPorowski) [inline docs/architecture/terraform-bicep-config.md]: What if `.terraformrc` / `terraform.rc` already exists (provided by user)?

File: docs/architecture/terraform-bicep-config.md [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+### Example usage
+
+```bicep
+resource tfConfig 'Radius.Core/terraformConfigs@2025-08-01-preview' = {
+  name: 'corp-terraform'
+  properties: {
+    terraformrc: {
+      providerInstallation: {
+        networkMirror: {
```
Comment (willdavsmith) [inline docs/architecture/terraform-bicep-config.md:208]: do we support manual binary upload or only network download

File: docs/architecture/terraform-bicep-config.md [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+          url: 'https://mirror.corp.example.com/terraform/providers'
+          include: ['*']
+          exclude: ['hashicorp/azurerm']
```
Comment (willdavsmith) [inline docs/architecture/terraform-bicep-config.md:211]: What is the exclude key?

File: docs/architecture/terraform-bicep-config.md [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+          secret: tfcTokenSecret.id
+        }
+      }
+    }
+    env: {
+      TF_LOG: 'WARN'
+      TF_REGISTRY_CLIENT_TIMEOUT: '15'
+    }
+  }
+}
```
Comment (willdavsmith) [inline docs/architecture/terraform-bicep-config.md:228]: if a user doesn't provide this, do we have an auto-configured terraform install (like we do today)?

File: docs/architecture/terraform-bicep-config.md [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+
+resource bicepConfig 'Radius.Core/bicepConfigs@2025-08-01-preview' = {
+  name: 'corp-bicep'
+  properties: {
+    registryAuthentication: {
+      authenticationMethod: 'BasicAuth'
+      basicAuthSecretId: acrSecret.id
+    }
```
Comment (willdavsmith) [inline docs/architecture/terraform-bicep-config.md:238]: do we need to handle mirrored bicep download here?

File: docs/architecture/terraform-bicep-config.md (hunk 3 above)
Comment (sylvainsf) [reply docs/architecture/terraform-bicep-config.md:208]: Out of scope for this change, this is the network mirror for terraform providers not the binary.

File: docs/architecture/terraform-bicep-config.md (hunk 4 above)
Comment (sylvainsf) [reply docs/architecture/terraform-bicep-config.md:211]: "don't use local mirror for these providers"

File: pkg/corerp/frontend/controller/environments/v20250801preview/createorupdateenvironment.go [hunk 7]
```diff
@@ -149,3 +162,31 @@ func (e *CreateOrUpdateEnvironmentv20250801preview) validateRecipePacks(ctx cont
...
+	id, parseErr := resources.Parse(resourceID)
+	if parseErr != nil {
+		return rest.NewBadRequestResponse(fmt.Sprintf("Invalid %s resource ID: %s", propertyName, resourceID))
+	}
+	if !strings.EqualFold(id.Type(), expectedType) {
+		return rest.NewBadRequestResponse(fmt.Sprintf("Referenced %s resource %q has type %q; expected %q.", propertyName, resourceID, id.Type(), expectedType))
+	}
+	if _, _, err := e.GetResource(ctx, id); err != nil {
+		return rest.NewBadRequestResponse(fmt.Sprintf("Referenced %s resource %q does not exist.", propertyName, resourceID))
+	}
```
Comment (brooke-hamilton) [inline pkg/corerp/frontend/controller/environments/v20250801preview/createorupdateenvironment.go:202]: **Bug: this existence check never fires.**

`Operation[P,T].GetResource` (`pkg/armrpc/frontend/controller/operation.go:104-107`) explicitly swallows `database.ErrNotFound` and returns `(nil, "", nil)`. I verified by calling `inmemory.Get` for a missing ID — it returns `*database.ErrNotFound`, which `errors.Is` matches, so `Operation.GetResource` clears `err` before returning.

As written, a typo'd or deleted `terraformConfig` / `bicepConfig` ID passes PUT validation silently and only fails much later at recipe-execution time — exactly the regression the doc-comment on `validateConfigRef` says it prevents.

Suggested fix:

```go
out, _, err := e.GetResource(ctx, id)
if err != nil {
    return rest.NewInternalServerErrorARMResponse(...) // or propagate
}
if out == nil {
    return rest.NewBadRequestResponse(fmt.Sprintf("Referenced %s resource %q does not exist.", propertyName, resourceID))
}
```

The preceding `id.Type()` equality check on line 185 already enforces the type, so existence is the only thing the Get needs to prove. A unit test for `validateConfigRef` with a non-existent ID would lock this in.

File: pkg/recipes/terraform/execute.go [hunk 8]
```diff
@@ -227,6 +233,25 @@ func (e executor) setEnvironmentVariables(tf *tfexec.Terraform, options Options)
...
+			options.Secrets,
+		)
+		if err != nil {
+			return err
+		}
+		if rcPath != "" {
+			envVarUpdate = true
+			envVars[envTFCLIConfigFile] = rcPath
+		}
+	}
```
Comment (brooke-hamilton) [inline pkg/recipes/terraform/execute.go:256]: **Bug: module download runs before `TF_CLI_CONFIG_FILE` is set.**

In `Deploy` the call order is `generateConfig` (line 79) → `setEnvironmentVariables` (line 86). `generateConfig` calls `downloadAndInspect` → `tf.Get(ctx)` (`module.go:67`), which fetches modules. For module sources hosted on a Terraform registry (e.g. `app.terraform.io/foo/bar/aws`) that requires a `credentials "host" {}` block, those credentials are not yet in effect when `tf.Get` runs — only later when `tf.Init` downloads providers.

Since the PR description explicitly markets `terraformrc.credentials` as registry HTTP auth, this gap is worth fixing now. Hoisting `applyTerraformCLIConfig` (or the `.terraformrc`-writing portion of `setEnvironmentVariables`) ahead of `generateConfig` would close it.

A related concern: `applyTerraformCLIConfig` is the only env setup `Delete` does. Env vars and secrets from `recipeConfig.Env.AdditionalProperties` / `EnvSecrets` are not propagated to `terraform destroy`. Consider extracting a shared helper that always applies env vars, secrets, and the rc-file env in both Deploy and Delete.

File: pkg/corerp/api/v20250801preview/terraformconfig_conversion.go [hunk 9]
```diff
@@ -0,0 +1,155 @@
+/*
```
Comment (brooke-hamilton) [inline pkg/corerp/api/v20250801preview/terraformconfig_conversion.go:1]: **Test coverage gap: missing `terraformconfig_conversion_test.go`.**

Repo convention is that every `*_conversion.go` has a sibling `*_conversion_test.go` — `environment_conversion_test.go`, `application_conversion_test.go`, and `recipepack_conversion_test.go` all exist. Add round-trip tests covering at minimum:

- Nil `Terraformrc`, nil `ProviderInstallation`, nil `Credentials` (no panics).
- `NetworkMirror` only / `Direct` only / both.
- `Credentials` map with multiple hosts.
- Round-trip identity (`ConvertTo` then `ConvertFrom` produces equal data).

File: pkg/corerp/api/v20250801preview/bicepconfig_conversion.go [hunk 10]
```diff
@@ -0,0 +1,126 @@
+/*
```
Comment (brooke-hamilton) [inline pkg/corerp/api/v20250801preview/bicepconfig_conversion.go:1]: **Test coverage gap: missing `bicepconfig_conversion_test.go`.**

Same convention as the other versioned converters in this directory. Worth covering:

- All three `authenticationMethod` values (`BasicAuth`, `AzureWI`, `AwsIrsa`).
- Empty `RegistryAuthentications` map.
- Round-trip identity.
- The aliasing pattern in `fromBicepRegistryAuthDataModel` (lines 113-124) takes the address of fields on the function parameter — a test that builds two distinct entries in the same map would catch any future regression where pointer aliasing causes shared state.

File: pkg/corerp/frontend/controller/environments/v20250801preview/createorupdateenvironment.go [hunk 11]
```diff
@@ -103,6 +104,18 @@ func (e *CreateOrUpdateEnvironmentv20250801preview) Run(ctx context.Context, w h
...
+	if newResource.Properties.TerraformConfig != "" {
+		if resp := validateConfigRef(ctx, e, newResource.Properties.TerraformConfig, datamodel.TerraformConfigResourceType, "terraformConfig"); resp != nil {
+			return resp, nil
+		}
+	}
+	if newResource.Properties.BicepConfig != "" {
+		if resp := validateConfigRef(ctx, e, newResource.Properties.BicepConfig, datamodel.BicepConfigResourceType, "bicepConfig"); resp != nil {
+			return resp, nil
+		}
+	}
```
Comment (brooke-hamilton) [inline pkg/corerp/frontend/controller/environments/v20250801preview/createorupdateenvironment.go:117]: **Test coverage gap: no controller tests for the new `terraformConfig` / `bicepConfig` validation.**

`createorupdateenvironment_test.go` has zero references to `validateConfigRef`, `TerraformConfig`, or `BicepConfig`. Cases worth adding:

- Invalid resource ID (parse failure → 400).
- Wrong resource type (e.g. environment ID supplied as `terraformConfig` → 400).
- Non-existent resource of the right type → 400. **(This is the test that would have caught the existence-check bug above.)**
- Happy path: existing `terraformConfig` and `bicepConfig` accepted.

**Review by brooke-hamilton — COMMENTED** [review]
Minor issues only.

File: pkg/recipes/configloader/environment.go [hunk 12]
```diff
@@ -190,6 +190,85 @@ func getConfigurationV20250801(environment *v20250801preview.EnvironmentResource
...
+					Secret: auth.BasicAuthSecretId,
+				}
+			}
+			if len(authMap) > 0 {
+				config.RecipeConfig.Bicep = datamodel.BicepConfigProperties{
+					Authentication: authMap,
+				}
+			}
+		}
+	}
```
Comment (brooke-hamilton) [inline pkg/recipes/configloader/environment.go:270]: **Test coverage gap: no unit tests for `getConfigurationV20250801`'s new bridge logic.**

The functional test exercises the happy path end-to-end, but unit tests would lock in the mapping rules and surface regressions much faster:

- `tfProps.Terraformrc.Credentials` are copied into `RecipeConfig.Terraform.Credentials`.
- `tfProps.Env` is mapped to `RecipeConfig.Env.AdditionalProperties`.
- `tfProps.Terraformrc.ProviderInstallation` pointer is forwarded as-is.
- For BicepConfig: entries with empty `BasicAuthSecretId` are skipped (currently silent — a unit test pins this behavior).
- Errors from `FetchTerraformConfig` / `FetchBicepConfig` are wrapped with the resource ID.

**Comment by sylvainsf** [comment]
Filed #11856 for the Delete env/secrets asymmetry — leaving it out of this PR to keep the scope focused.

**Review by brooke-hamilton — APPROVED** [review]
🚀
